gdb/uftrace/rbtree.py

```diff
@@ -17,6 +17,115 @@
 rb_root = utils.CachedType("struct rb_root")
 rb_node = utils.CachedType("struct rb_node")
 
+def rb_color(node):
+    """Return the color of a node.
+    red -> 0 | black -> 1"""
+    if node.address == 0:
+        return 1
+
+    if node['rb_parent_color'] % 2 == 0:
+        return 0
+    else:
+        return 1
+
+def rb_check(node):
+    if node.address == 0:
+        return 1
+
+    left = node['rb_left'].dereference()
+    right = node['rb_right'].dereference()
+
+    # check that a red node has black children
+    if rb_color(node) == 0:
+        if rb_color(left) == 0:
+            gdb.write(f"red node {node} has red left child {left}\n")
+            return -1
+        if rb_color(right) == 0:
+            gdb.write(f"red node {node} has red right child {right}\n")
+            return -1
+
+    # recursively check that paths to NIL have as many black nodes
+    left_black_count = rb_check(left)
+    if left_black_count == -1:
+        return -1
+
+    right_black_count = rb_check(right)
+    if right_black_count == -1:
+        return -1
+
+    if left_black_count != right_black_count:
+        gdb.write(f"node {node.address}: {left_black_count} on left != {right_black_count} on right\n")
+        return -1
+    else:
+        black_count = left_black_count
+        if rb_color(node) == 1:
+            black_count += 1
+        return black_count
+
+
+class RbTreeCheck(gdb.Command):
+    """Check is a rbtree is valid."""
```

Comment:
Can you please describe what to check briefly?  I almost forgot about the detail of the RB tree constraints :-p